sql/catalyst/src/main/scala/org/apache/spark/sql/catalog/v2/expressions/expressions.scala

```diff
@@ -0,0 +1,194 @@
+    literal(value, internalLit.dataType)
+  }
+
+  def literal[T](value: T, dataType: DataType): LiteralValue[T] = LiteralValue(value, dataType)
+
+  def reference(name: String): NamedReference =
+    FieldReference(parser.parseMultipartIdentifier(name).toArray)
+
+  def apply(name: String, arguments: Array[Expression]): Transform = ApplyTransform(name, arguments)
+
+  def apply(name: String, arguments: Expression*): Transform = ApplyTransform(name, arguments)
+
+  def bucket(numBuckets: Int, columns: Array[String]): BucketTransform =
+    BucketTransform(literal(numBuckets, IntegerType), columns.map(reference))
+
+  def bucket(numBuckets: Int, columns: String*): BucketTransform =
+    BucketTransform(literal(numBuckets, IntegerType), columns.map(reference))
+
+  def identity(column: String): IdentityTransform = IdentityTransform(reference(column))
+
+  def year(column: String): YearTransform = YearTransform(reference(column))
+
+  def month(column: String): MonthTransform = MonthTransform(reference(column))
+
+  def date(column: String): DateTransform = DateTransform(reference(column))
+
+  def dateHour(column: String): DateHourTransform = DateHourTransform(reference(column))
+}
+
+/**
+ * Base class for simple transforms of a single column.
+ */
+private[sql] abstract class SingleColumnTransform(ref: NamedReference) extends Transform {
+
+  def reference: NamedReference = ref
+
+  override lazy val references: Array[NamedReference] = Array(ref)
+
+  override lazy val arguments: Array[Expression] = Array(ref)
+
+  override lazy val describe: String = name + "(" + reference.describe + ")"
+
+  override def toString: String = describe
+}
+
+private[sql] final case class BucketTransform(
+    numBuckets: Literal[Int],
+    columns: Seq[NamedReference]) extends Transform {
+
+  override lazy val name: String = "bucket"
+
+  override lazy val references: Array[NamedReference] = {
+    arguments
+        .filter(_.isInstanceOf[NamedReference])
+        .map(_.asInstanceOf[NamedReference])
+  }
+
+  override lazy val arguments: Array[Expression] = numBuckets +: columns.toArray
+
+  override lazy val describe: String = s"bucket(${arguments.map(_.describe).mkString(", ")})"
+
+  override def toString: String = describe
+}
+
+private[sql] final case class ApplyTransform(
+    name: String,
+    args: Seq[Expression]) extends Transform {
+
+  override lazy val arguments: Array[Expression] = args.toArray
+
+  override lazy val references: Array[NamedReference] = {
+    arguments
+        .filter(_.isInstanceOf[NamedReference])
+        .map(_.asInstanceOf[NamedReference])
+  }
+
+  override lazy val describe: String = s"$name(${arguments.map(_.describe).mkString(", ")})"
+
+  override def toString: String = describe
+}
+
+private[sql] final case class IdentityTransform(
+    ref: NamedReference) extends SingleColumnTransform(ref) {
+  override lazy val name: String = "identity"
+  override lazy val describe: String = ref.describe
```

Comment:
we should not abuse `lazy val`. `def` is better here.

Reply:
Please be more clear. Why is using a `lazy val` an abuse?

Reply:
`lazy val` has overhead, if we check the generated java bytecode, `lazy val` needs to add extra member variables to the class, and needs to take care of multi-thread, etc.

A simple method returning a string constant does not worth the overhead of `lazy val`.

Reply:
I see what you mean about some overhead, but I don't think this is an abuse. This may look like one because it is a lazy val referencing the value of another lazy val, but this must be a lazy val because it is a lazy the parent class.

Most of the `describe` implementations are lazy to avoid repeated string manipulation when we can just keep track of the computed value.

Reply:
`lazy val` has much more overhead than string manipulation, AFAIK. We should avoid using `lazy val` in the parent class as well.

If we call `describe` a million times per second, then maybe `lazy val` can bring some benefits like less GC. But I don't think this is the case here.

Reply:
BTW, we shouldn't consider performance at this stage, and using `def` is more natural here.

Reply:
I agree that we don't really need to care about performance here, but I think `lazy val` is more natural.

Reply:
If you check the `toString` and `sql` methods in various expressions, logical plans, phyiscal plans, most of them use `def` instead of `lazy val`.

Maybe we can't reach an agreement about which one is more natural, but `def` is definitely more consistent with the Spark code base.

Reply:
Is this important to change, or is this a preference?

Reply:
I think it's more than a preference. PRs should keep the code style consistent with the code base.